naptime/src/main/scala/org/coursera/naptime/ari/engine/Utilities.scala

```diff
@@ -36,51 +36,94 @@ import scala.collection.JavaConverters._
 
 
 object Utilities extends StrictLogging {
-
-  // TODO(bryan): Dedupe this from NaptimeUnionField
-  val TYPED_DEFINITION_KEY = "typedDefinition"
-  def getTypedDefinition(dataSchema: DataSchema): Option[Map[String, String]] = {
-    Option(dataSchema).flatMap(schema => Option(schema.getProperties.get(TYPED_DEFINITION_KEY)).collect {
-      case definitions: java.util.Map[String @unchecked, String @unchecked] => definitions.asScala.toMap
-    })
-  }
-
-  private[this] def replaceTypedDefinitionNames(dataElement: DataElement): List[String] = {
-    val thisName = (for {
-      parent <- Option(dataElement.getParent)
-      schema <- Option(parent.getSchema)
-      _ <- getTypedDefinition(schema)
-        if dataElement.getName.toString == "definition"
-      typeName <- Option(dataElement.getParent.getValue).flatMap {
-        case dataMap: DataMap => Option(dataMap.getString("typeName"))
-        case _ => None
-      }
-    } yield {
-      typeName
-    }).getOrElse(dataElement.getName.toString)
-    val parentNames = Option(dataElement.getParent).map(replaceTypedDefinitionNames)
-    (parentNames.getOrElse(List.empty) :+ thisName).filterNot(_.isEmpty)
+  private val TYPED_DEFINITION_KEY = "typedDefinition"
+  private def getTypedDefinitionMappings(elements: Iterator[DataElement]): Map[String, String] = {
+    val typedDefinitionMappings = elements.flatMap { element =>
+      Option(element.getSchema.getProperties.get(TYPED_DEFINITION_KEY)).collect {
+        case definitions: java.util.Map[String@unchecked, String@unchecked] => definitions.asScala.toList
```

Comment:
i might be missing something but is there something more efficient we can do here? it seems that we're converting the java map to a scala list only to return a map again?

Reply:
There is a potential Java map for each element (`element.getSchema.getProperties.get(TYPED_DEFINITION_KEY))`) which we use `.collect` to grab and turn into a list of maps which is then turned into a single map. 

Let me get rid of the `.toList` part, we don't need to convert to a Scala list. I also added an intermediate variable for clarity.